### PR TITLE
Branch Policy

### PR DESCRIPTION
**Pull request by t8m**
This is a draft of the branch policy for starting the discussion. There are two additional branches - a **future major branch** and a **future minor branch**.

This should allow for merging PRs that would destabilize development of the next release that is currently in works. I.E. these two branches provide a staging area for patches that we want in future releases being it either a major or minor release based on whether the change involves API breakage or not.

### CODE REVIEW

File: policies/branch-policy.md
Comment (kroeckx) [inline policies/branch-policy.md:16]: It's unclear to me what you mean with approval of a branch. Do you mean release?

File: policies/branch-policy.md
Comment (kroeckx) [inline policies/branch-policy.md:18]: Do you really mean major here?

File: policies/branch-policy.md
Comment (t8m) [reply policies/branch-policy.md:16]: Basically this PR consists of two proposals for further discussion. It is still in kind-of draft state. One proposal is to have a feature branch and the other is to have a next-minor branch to develop the minor release keeping the master branch for larger refactorings/API changes.

File: policies/branch-policy.md
Comment (t8m) [reply policies/branch-policy.md:18]: Yes, we are talking here about what is the purpose of the master branch.

File: policies/branch-policy.md
Comment (kaduk) [inline policies/branch-policy.md:14]: I think that having extended periods of time where API/ABI-breaking changes are not allowed introduces a large risk of slowing down development.  If large projects/refactorings might be stalled for years (i.e., if next-minor is not approved), there is reduced incentive to start on them and incur the ongoing maintenance burden to keep them up to date until they can be reviewed and merged.

We already have numerous data points of submissions from the community of large features that have been awaiting review for literally years, with no sign of such review arriving anytime soon, either.  Please consider the effects of this lack of motion on the broader community.

File: policies/branch-policy.md
Comment (kaduk) [inline policies/branch-policy.md:37]: Is there to be any distinction between release branches for LTS and non-LTS release branches?
(I personally am okay with no difference, but want to check if that is the intent.)

File: policies/branch-policy.md
Comment (kaduk) [inline policies/branch-policy.md:73]: Having the PR submitter indicate their desire seems quite useful.
I am not sure that trying to say "minor refactoring is allowed" will help much, since the boundary between minor/major is subjective.

File: policies/branch-policy.md
Comment (kaduk) [inline policies/branch-policy.md:95]: I am not sure that actual branch renaming will interact nicely with git clients.
Wouldn't the second time we use `next-minor` incur a non-fast-forward update to the branch?

File: policies/branch-policy.md
Comment (t8m) [reply policies/branch-policy.md:37]: In terms of what is allowed to commit to LTS and non-LTS release branches this policy does not distinguish them. We might modify the policy later if we find that LTS branches need some different treatment.

File: policies/branch-policy.md
Comment (t8m) [reply policies/branch-policy.md:73]: I agree. I do not like these subjective adverbs in policies. And it would be really hard to give some exact limits such as number of changed lines or so because renaming an internal function might give pretty large PR but without any serious destabilization impact.

File: policies/branch-policy.md
Comment (t8m) [reply policies/branch-policy.md:95]: Yeah it would I suppose. In that case we might name the branch openssl-x.y from start. I just wanted to give an option to the OMC to decide to use some different y version although that seems not too important. Also renaming (or simply re-branchng) openssl-3.1 to openssl-3.13 if OMC decided that the next version will have that minor number seems less problematic.

Comment (t8m) [reply policies/branch-policy.md:95]: I'll modify the proposal.

File: policies/branch-policy.md
Comment (levitte) [inline policies/branch-policy.md:6]: Nit: `##` should be `###`, because you have a level two title further down (underlined with `--------...`)

File: policies/branch-policy.md
Comment (levitte) [inline policies/branch-policy.md:76]: ... alternatively, change `----------------------------` to `============================`

File: policies/branch-policy.md
Comment (levitte) [inline policies/branch-policy.md:15]: I would drop "forward" here

**Review by levitte — COMMENTED** [review]
Why *next* major but *future* minor?
Aren't they both *next*? Or *future* for that matter?

**Comment by t8m** [comment]
> Why _next_ major but _future_ minor?
> Aren't they both _next_? Or _future_ for that matter?

Because the future minor branch is for changes that should go into the release that will be developed after the one we are currently developing is released. So that is a release after the next minor release.

The next major is a branch for changes that should go into the next major release.

I could possibly go with future major and future minor because the future major is for a future release. I would not call the future minor next minor as that would be IMO confusing.

If you have better names for these please suggest.

File: policies/branch-policy.md
Comment (t8m) [reply policies/branch-policy.md]: Done

**Comment by levitte** [comment]
> If you have better names for these please suggest.

Simply pick one and stick with it.  In this context, I'm reading "next" and "future" as synonyms

File: policies/branch-policy.md
Comment (mattcaswell) [inline policies/branch-policy.md:44]: In this case do we really need this branch? It seems redundant in this scenario

File: policies/branch-policy.md
Comment (mattcaswell) [inline policies/branch-policy.md:23]: Should we simply refer to the stable release updates policy here?

**Review by mattcaswell — CHANGES_REQUESTED** [review]

File: policies/branch-policy.md
Comment (t8m) [reply policies/branch-policy.md:23]: I can add a link.

File: policies/branch-policy.md
Comment (t8m) [reply policies/branch-policy.md:44]: It's true that we would have to somehow keep it in sync with the master branch. However it makes the source git more predictable to outside contributors I think.

Comment (levitte) [reply policies/branch-policy.md:44]: Hmm, depends...  as long as we do like now, with master being the development of the next release, major or minor alike, then I don't see the point.  What this suggests is that we would have multiple development branches...  and I can't recall how well we've discussed that.

Comment (t8m) [reply policies/branch-policy.md:44]: > What this suggests is that we would have multiple development branches... and I can't recall how well we've discussed that.

We discussed it that we want it the last time we discussed the policy proposal on OTC meeting. So we can merge stuff to the future-major or future-minor branches and not keep open PRs for indefinite amount of time. But yeah, further discussion is needed for sure.

File: policies/branch-policy.md
Comment (romen) [inline policies/branch-policy.md:27]: Should this exception be decided by OTC?

File: policies/branch-policy.md
Comment (romen) [inline policies/branch-policy.md:60]: Why would we need a future minor, when the future release is scheduled to be a major release?

File: policies/branch-policy.md
Comment (romen) [reply policies/branch-policy.md:43]: IIRC the need for a future-major branch when the next release is going to be a major release is supposed to exist for those PRs that OMC decides are out of scope for the next release and are deferred to the following major release.

In this case this item needs to be rewritten, as this branch and master might be out of sync in such a scenario.

File: policies/branch-policy.md
Comment (t8m) [reply policies/branch-policy.md:27]: IMO OTC can give an exception only if some change is necessary for fixing a bug even if that change by itself could be seen as a feature.

File: policies/branch-policy.md
Comment (t8m) [reply policies/branch-policy.md:60]: I'll try to make this a little bit more general.

File: policies/branch-policy.md
Comment (slontis) [inline policies/branch-policy.md:16]: Should this implicitly say changes need to be backported to the supported release branches also.. (It is currently implied since changes to the future branches are backported).

**Review by slontis — COMMENTED** [review]
Does this need to mention master at all (or the lack of master)

**Comment by slontis** [comment]
So where are we currently if we were doing this now?
master would somehow disappear.
3.1 is the dev branch and 4.0 is the future major with no future minor?
After 3.1, there would either be a 3.2 or 4.0 as the dev branch?

**Comment by t8m** [comment]
> 3.1 is the dev branch and 4.0 is the future major with no future minor?

Unless OMC would be planning to release 4.0 after 3.1 we would actually have 3.2 as the future minor branch (if we had anything to put there actually).

**Comment by t8m** [comment]
OTC: Add requirement to approve merges of PRs to future minor and future major branch by explicit OTC approval (by consensus or a vote).

File: policies/branch-policy.md
Comment (mattcaswell) [inline policies/branch-policy.md:14]: Presumably whether or not it is a major or minor is implicit in the name, e.g. is it openssl-3.1 or openssl-4.0. We're not planning on changing the name after creation right?

File: policies/branch-policy.md
Comment (mattcaswell) [inline policies/branch-policy.md:23]: Perhaps supported stable releases instead?

**Review by mattcaswell — CHANGES_REQUESTED** [review]

File: policies/branch-policy.md
Comment (t8m) [reply policies/branch-policy.md:23]: Yep

File: policies/branch-policy.md
Comment (t8m) [reply policies/branch-policy.md]: Yes, I've simplified the text here.

**Comment by t8m** [comment]
Added the approval for merges to future branches by OTC.

File: policies/branch-policy.md
Comment (paulidale) [inline policies/branch-policy.md:11]: Does this mean we cut releases from the development branch despite having future major and minor branches?

**Comment by fariborz4** [comment]
Y

File: policies/branch-policy.md
Comment (fariborz4) [reply policies/branch-policy.md]: H

File: policies/branch-policy.md
Comment (fariborz4) [reply policies/branch-policy.md]: H

File: policies/branch-policy.md
Comment (fariborz4) [reply policies/branch-policy.md:13]: H

File: policies/branch-policy.md
Comment (t8m) [reply policies/branch-policy.md:11]: Yes, and after the (final or perhaps beta?) release is cut a future minor or a future major (depending on the next release type) will become the default development branch. That's why the name of the branch never changes and is `openssl-x.y`.

**Comment by t8m** [comment]
@fariborz4 please stop producing this kind of spam comments, otherwise we would have to block you.

File: policies/branch-policy.md
Comment (t8m) [reply policies/branch-policy.md:13]: IMO no, because this applies only to bug fixes. And backporting bug fixes is handled on line 28.

**Comment by t8m** [comment]
Starting the vote for Add Branch Policy at commit [10bea43](https://github.com/openssl/technical-policies/pull/7/commits/10bea43ee89eb05949464f3c50b06d864991862f)

**Comment by mattcaswell** [comment]
Vote: [+1]

**Comment by romen** [comment]
Vote: [+1]

**Comment by slontis** [comment]
Vote: [+0]

**Comment by t-j-h** [comment]
Vote: [+1]

I also note that the description of the types of branches would indicate that the planned OpenSSL-3.1 release should really be called OpenSSL-4.0 (given it is an implementation of a new protocol). I don't think the descriptions should change.

**Comment by paulidale** [comment]
Vote: [0]

**Comment by kaduk** [comment]
The actual text has gotten to a pretty significant level of abstraction, so I just want to check my understanding: the "default development branch" is always going to be of the form openssl-x.y for some 'x' and 'y', with the holder of the "default" label changing when a release is cut from it?  I.e., there will not be a branch named "master" or "main" or whatnot?

**Comment by kaduk** [comment]
The actual text has gotten to a pretty significant level of abstraction, so I just want to check my understanding: the "default development branch" is always going to be of the form openssl-x.y for some 'x' and 'y', with the holder of the "default" label changing when a release is cut from it?  I.e., there will not be a branch named "master" or "main" or whatnot?

**Comment by slontis** [comment]
I also asked this question initially.

> The actual text has gotten to a pretty significant level of abstraction, so I just want to check my understanding: the "default development branch" is always going to be of the form openssl-x.y for some 'x' and 'y', with the holder of the "default" label changing when a release is cut from it? I.e., there will not be a branch named "master" or "main" or whatnot?

**Comment by mattcaswell** [comment]
> The actual text has gotten to a pretty significant level of abstraction, so I just want to check my understanding: the "default development branch" is always going to be of the form openssl-x.y for some 'x' and 'y', with the holder of the "default" label changing when a release is cut from it? I.e., there will not be a branch named "master" or "main" or whatnot?

Correct

**Comment by t8m** [comment]
> I also note that the description of the types of branches would indicate that the planned OpenSSL-3.1 release should really be called OpenSSL-4.0 (given it is an implementation of a new protocol). I don't think the descriptions should change.

I do not think the interpretation must be so strict. What's described here is a general intention but OMC has always the last word on setting the release versions. Also, as 3.1 will not have a full-fledged implementation of QUIC, we might want to reserve 4.0 for the release in which we will regard QUIC implementation reasonably complete.

**Comment by beldmit** [comment]
Vote: [0]

File: policies/branch-policy.md
Comment (levitte) [inline policies/branch-policy.md:29]: If someone reports an error in 1.1.1, I usually make the fix against 1.1.1, not master.  Does this section contradict that?  I'm a little unsure how to interpret the wordings for that kind of scenario.

Comment (levitte) [reply policies/branch-policy.md:29]: I'll wait for the answer before voting

**Comment by levitte** [comment]
> Vote: [+1]
> 
> I also note that the description of the types of branches would indicate that the planned OpenSSL-3.1 release should really be called OpenSSL-4.0 (given it is an implementation of a new protocol). I don't think the descriptions should change.

In the minor release branch section, there's this:

``` markdown
- No major features are allowed unless explicitly acked by OMC as targeted for
  the minor release.
```

QUIC MVP has been actively proposed by the OMC to be for OpenSSL 3.1, so I think we're good :wink:

File: policies/branch-policy.md
Comment (t8m) [reply policies/branch-policy.md:29]: IMO it does not preclude that, if you ensure that the fix (if the bug applies to 3.0 and master) applies to later branches. If it does not apply cleanly you'll need to create the PR against these branches as well. This talks about merging, i.e. no fix should be merged to 1.1.1 (if the bug is on 3.0 as well) without merging also to 3.0 and master first (or at least immediately after).

**Comment by t8m** [comment]
@kroeckx @levitte @mspncp ping for votes

**Comment by mspncp** [comment]
Vote: [+1]

File: policies/branch-policy.md
Comment (levitte) [reply policies/branch-policy.md:29]: > This talks about merging, i.e. no fix should be merged to 1.1.1 (if the bug is on 3.0 as well) without merging also to 3.0 and master first (or at least immediately after).

Agreed, and that's how I usually do things

File: policies/branch-policy.md
Comment (kroeckx) [inline policies/branch-policy.md:74]: Does this mean we're going to rename master to openssl-3.1?

**Comment by kroeckx** [comment]
Voting -1

File: policies/branch-policy.md
Comment (mattcaswell) [reply policies/branch-policy.md:74]: Yes

**Comment by t8m** [comment]
@levitte last chance for voting

**Comment by levitte** [comment]
Vote: [+0]

**Comment by t8m** [comment]
Closing the vote.
```
Topic: Add Branch Policy at commit 10bea43
       This will become an official OTC policy.
Proposed by: Tomas
Issue link: https://github.com/openssl/technical-policies/pull/7
Public: yes
Opened: 2022-05-25
Closed: 2022-06-08
Accepted: yes   (for: 5, against: 1, abstained: 4, not voted: 0)

   Dmitry     [ 0]
   Matt       [+1]
   Pauli      [ 0]
   Tim        [+1]
   Richard    [+0]
   Shane      [+0]
   Tomas      [+1]
   Kurt       [-1]
   Matthias   [+1]
   Nicola     [+1]
```

**Pull request closed without merging**
